{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "edb50bd7-0a0d-4464-b495-cfd722ef8b6f",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/matthew.schmitz/Matthew/utils/miniforge3/envs/rapids-24.12/lib/python3.12/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html\n",
      "  from .autonotebook import tqdm as notebook_tqdm\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GPU is not available\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import inspect\n",
    "import seaborn\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "import torch\n",
    "import scanpy as sc\n",
    "import tqdm\n",
    "import sys\n",
    "import pickle\n",
    "import PyComplexHeatmap as pch\n",
    "import scvi\n",
    "import IPython\n",
    "import scipy\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import anndata\n",
    "import gseapy\n",
    "import statannotations\n",
    "from statannotations.Annotator import Annotator\n",
    "\n",
    "sc.settings.figdir=os.path.expanduser('~/WbFigures/DevAnalysis')\n",
    "sc._settings.settings._vector_friendly=True\n",
    "\n",
    "device = torch.device(\"cuda:0\" if torch.cuda.is_available() else \"cpu\")\n",
    "\n",
    "if torch.cuda.is_available():\n",
    "    print(\"GPU is available\")\n",
    "    print(\"Number of GPUs:\", torch.cuda.device_count())\n",
    "    print(\"GPU Name:\", torch.cuda.get_device_name(0))\n",
    "else:\n",
    "    print(\"GPU is not available\")\n",
    "\n",
    "sys.path.append('/home/matthew.schmitz/utils/mts-utils/')\n",
    "from genomics import sc_analysis\n",
    "\n",
    "\n",
    "import sys\n",
    "import antipode\n",
    "import antipode.antipode_model\n",
    "import antipode.model_functions\n",
    "import antipode.model_distributions\n",
    "import antipode.model_modules\n",
    "import antipode.train_utils\n",
    "import antipode.plotting\n",
    "import antipode.post\n",
    "\n",
    "from antipode.post import uniqlist"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "e676dbd1-79ba-4101-a55b-c22427b43fb6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Pretty printing has been turned OFF\n"
     ]
    }
   ],
   "source": [
    "%pprint"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "f16e32a7-e5e5-4ebd-9e00-b52e55bed1e0",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[34mINFO    \u001b[0m Generating sequential column names                                                                        \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/matthew.schmitz/Matthew/utils/miniforge3/envs/rapids-24.12/lib/python3.12/site-packages/pyro/params/param_store.py:334: FutureWarning: You are using `torch.load` with `weights_only=False` (the current default value), which uses the default pickle module implicitly. It is possible to construct malicious pickle data which will execute arbitrary code during unpickling (See https://github.com/pytorch/pytorch/blob/main/SECURITY.md#untrusted-models for more details). In a future release, the default value for `weights_only` will be flipped to `True`. This limits the functions that could be executed during unpickling. Arbitrary objects will no longer be allowed to be loaded via this mode unless they are explicitly allowlisted by the user via `torch.serialization.add_safe_globals`. We recommend you start setting `weights_only=True` for any use case where you don't have full control of the loaded file. Please open an issue on GitHub for any issues related to this experimental feature.\n",
      "  state = torch.load(input_file, map_location)\n"
     ]
    }
   ],
   "source": [
    "batch_key='batch_name'\n",
    "discov_key='species'\n",
    "layer_key='spliced'\n",
    "MDE_KEY = \"X_antipode_UMAP\"\n",
    "leaf_key = 'level_2'\n",
    "anndata.settings.allow_write_nullable_strings = True\n",
    "\n",
    "model_tag = '1.9.1.8.5_Dev_final_noQ_INP_moredim_regionsample'\n",
    "model_path='/home/matthew.schmitz/Matthew/models/'+model_tag\n",
    "adata = sc.read_h5ad(os.path.join(model_path,'analysis_inprogress.h5ad'))\n",
    "antipode_model=antipode.antipode_model.ANTIPODE.load(model_path,adata=adata,prefix='p3_',device=device)\n",
    "antipode_model.save_params_to_uns()\n",
    "adata=antipode_model.adata_manager.adata\n",
    "pstore = adata.uns['param_store']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "ddc31373-1816-4dca-85cf-e8603668d91b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 1793824 × 16738\n",
       "    obs: 'G2M_score', 'S_score', 'batch_name', 'general_region', 'log10_n_counts', 'phase', 'region', 'species', 'female', 'general_region_species', '_scvi_discov_ind', '_scvi_batch_ind', 'psi_0', 'psi_1', 'psi_2', 'q_score', 'level_0', 'level_1', 'level_2', 'antipode_cluster', 'kmeans', 'anno_level_2', 'initial_class', 'ic_extended', 'abc_subclass_markercor', 'juv_subclass_markercor', 'linnarson_extended', 'lamanno_class', 'lamanno_subclass', 'lamanno_cluster', 'neighborhood', 'curated_class_label', 'new_ccl', 'new_neighborhood', 'ccl_extended', 'new_neighborhood_extended', 'n_genes', 'percent_ribo', 'msregion', 'singlet', 'individual', 'timepoint', 'n_counts', 'doublet', 'dataset_name', 'log_abs_sum_dynam', 'clipped_psi_2', 'log_n_genes'\n",
       "    var: 'highly_variable', 'highly_variable_intersection', 'highly_variable_nbatches', 'highly_variable_rank', 'mean', 'means', 'n_cells', 'std', 'variances', 'variances_norm'\n",
       "    uns: '_scvi_manager_uuid', '_scvi_uuid', 'antipode_cluster_colors', 'batch_name_colors', 'ccl_extended_colors', 'dendrogram_level_2', 'general_region_color_dict', 'general_region_colors', 'leiden', 'level_0_colors', 'level_1_color_dict', 'level_1_colors', 'level_2_color_dict', 'level_2_colors', 'neighborhood_color_dict', 'neighborhood_colors', 'new_neighborhood_extended_color_dict', 'new_neighborhood_extended_colors', 'param_store', 'pca', 'phase_color_dict', 'phase_colors', 'region_cats', 'region_color_dict', 'region_colors', 'region_colors_dict', 'species_color_dict', 'species_colors', 'umap'\n",
       "    obsm: 'X_antipode', 'X_antipode_UMAP', 'discov_onehot', 'kmeans_onehot', 'phase_sex', 'taxon_probs'\n",
       "    layers: 'spliced'"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "f94a8234-adb7-4155-a359-98afbf5b2518",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 696/696 [02:40<00:00,  4.33it/s]\n",
      "100%|██████████| 232/232 [01:47<00:00,  2.16it/s]\n"
     ]
    }
   ],
   "source": [
    "# aggr_means=antipode.model_functions.group_aggr_anndata(adata,[discov_key,'level_2'],layer=layer_key,normalize=True)\n",
    "# aggr_sums=antipode.model_functions.group_aggr_anndata(adata,['level_2'],layer=layer_key,normalize=False,agg_func=np.sum)\n",
    "# log_real_means=antipode.model_functions.safe_log_transform(aggr_means[0],aggr_sums[0].sum(-1)[np.newaxis,...,np.newaxis])\n",
    "log_real_means,real_mean_levels=antipode.model_functions.get_real_leaf_means(adata,discov_key,leaf_key,layer=layer_key)\n",
    "real_means=pd.DataFrame(log_real_means.mean(0),columns=adata.var.index,index=real_mean_levels[leaf_key])\n",
    "real_means=real_means.loc[real_mean_levels[leaf_key],:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "2df11bcc-5087-4aa5-a8bc-34d30d42de7f",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 201/201 [02:41<00:00,  1.25it/s]\n",
      "100%|██████████| 67/67 [01:45<00:00,  1.58s/it]\n"
     ]
    }
   ],
   "source": [
    "ccl_aggr_means=antipode.model_functions.group_aggr_anndata(adata,[discov_key,'ccl_extended'],layer=layer_key,normalize=True)\n",
    "ccl_aggr_sums=antipode.model_functions.group_aggr_anndata(adata,['ccl_extended'],layer=layer_key,normalize=False,agg_func=np.sum)\n",
    "ccl_log_real_means=antipode.model_functions.safe_log_transform(ccl_aggr_means[0],ccl_aggr_sums[0].sum(-1)[np.newaxis,...,np.newaxis])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "249a7c87-f7ec-472d-8aab-3eb4fb2775ff",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/allen/programs/celltypes/workgroups/rnaseqanalysis/EvoGen/Team/Matthew/code/scANTIPODE/antipode/antipode_mixins.py:445: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  zero_mask = (adata.obs.groupby(self.discov_key)[leaf_level].value_counts().unstack().loc[:,cluster_labels]>=cluster_count_threshold).to_numpy()\n",
      "/allen/programs/celltypes/workgroups/rnaseqanalysis/EvoGen/Team/Matthew/code/scANTIPODE/antipode/antipode_mixins.py:446: RuntimeWarning: divide by zero encountered in divide\n",
      "  return discov_cluster_params,cluster_params, cluster_labels,var_labels, 1/zero_mask[...,np.newaxis],(prop_taxon, prop_locs,prop_discov_di,prop_discov_dm)\n"
     ]
    }
   ],
   "source": [
    "pstore = adata.uns['param_store']\n",
    "discov_cluster_params,cluster_params, cluster_labels,var_labels, zero_mask,(prop_taxon, prop_locs,prop_discov_di,prop_discov_dm)=antipode_model.calculate_cluster_params(flavor='numpy')\n",
    "keep_clusters = np.array([x in adata.obs['level_2'].unique() for x in cluster_labels ])\n",
    "cluster_labels = np.array(cluster_labels)[keep_clusters]\n",
    "zero_mask = zero_mask[:,keep_clusters,:]\n",
    "cluster_index = np.array([int(x) for x in cluster_labels])\n",
    "zero_mask = np.nan_to_num(zero_mask,posinf=np.nan)\n",
    "discov_cluster_params = discov_cluster_params[:,cluster_index,:] * zero_mask\n",
    "prop_locs = prop_locs[cluster_index,:]\n",
    "prop_discov_di = prop_discov_di[:,cluster_index,:] * zero_mask\n",
    "prop_discov_dm = prop_discov_dm[:,cluster_index,:] * zero_mask\n",
    "discov_labels=adata.obs[antipode_model.discov_key].cat.categories\n",
    "latent_labels=[str(x) for x in range(pstore['discov_dc'].shape[1])]\n",
    "n_clusters=antipode_model.level_sizes[-1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "886eeb14-84c5-4d7f-bdc4-dbe83d28c76e",
   "metadata": {},
   "outputs": [],
   "source": [
    "clipped_expr_df = pd.DataFrame(np.clip(discov_cluster_params,a_min=-20,a_max=1e10).mean(0),columns=adata.var.index,index=cluster_labels)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "28651b88-b924-4b83-87a7-a487ae63e185",
   "metadata": {},
   "outputs": [],
   "source": [
    "quantile = 0.98\n",
    "coefdf = antipode.post.get_conserved_quantile_markers(discov_cluster_params,columns=adata.var.index,index=cluster_labels, q=quantile,antimarker=False)\n",
    "selected_markers = antipode.post.select_marker_genes(\n",
    "    coefdf, clipped_expr_df, marker=True,\n",
    "    threshold=np.log(5/1e6))\n",
    "marker_gene_order=uniqlist([selected_markers[x] for x in adata.uns['dendrogram_level_2']['categories_ordered']])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "435f020c-b61f-4d80-aa3a-5e4a42873feb",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 67/67 [03:42<00:00,  3.32s/it]\n"
     ]
    }
   ],
   "source": [
    "ccl_coefdf = antipode.post.get_conserved_quantile_markers(ccl_log_real_means,columns=adata.var.index,index=adata.obs['ccl_extended'].cat.categories, q=quantile,antimarker=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "418fc07a-db70-4125-9232-729decc7648f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>level_2</th>\n",
       "      <th>level_1</th>\n",
       "      <th>new_neighborhood_extended</th>\n",
       "      <th>ccl_extended</th>\n",
       "      <th>region</th>\n",
       "      <th>abc_subclass_markercor</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TAGTATGTAATA_e14-WT8</th>\n",
       "      <td>157</td>\n",
       "      <td>26</td>\n",
       "      <td>P</td>\n",
       "      <td>NPC</td>\n",
       "      <td>Ctx</td>\n",
       "      <td>320 Astro-OLF NN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTCGCATAGGTT_e14-WT8</th>\n",
       "      <td>271</td>\n",
       "      <td>11</td>\n",
       "      <td>T</td>\n",
       "      <td>Ctx_IT</td>\n",
       "      <td>Ctx</td>\n",
       "      <td>309 CB PLI Gly-Gaba</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CCTCCAATTGGC_e14-WT8</th>\n",
       "      <td>216</td>\n",
       "      <td>5</td>\n",
       "      <td>T</td>\n",
       "      <td>Ctx_Allo</td>\n",
       "      <td>Ctx</td>\n",
       "      <td>309 CB PLI Gly-Gaba</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CACGATATCCAT_e14-WT8</th>\n",
       "      <td>373</td>\n",
       "      <td>41</td>\n",
       "      <td>NN</td>\n",
       "      <td>Endothelial</td>\n",
       "      <td>Ctx</td>\n",
       "      <td>329 ABC NN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CAGCCGCGCCCT_e14-WT8</th>\n",
       "      <td>214</td>\n",
       "      <td>5</td>\n",
       "      <td>T</td>\n",
       "      <td>Ctx_Allo</td>\n",
       "      <td>Ctx</td>\n",
       "      <td>309 CB PLI Gly-Gaba</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGGTTTCCCTTGTG_SRR13565274_GSM5047783_GW21-04_cerebellum_Homo_sapiens_RNA-Seq_kOut</th>\n",
       "      <td>398</td>\n",
       "      <td>29</td>\n",
       "      <td>DMR</td>\n",
       "      <td>RE_ATOH1_Granule</td>\n",
       "      <td>Cb</td>\n",
       "      <td>314 CB Granule Glut</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGGTTTCTGCGGCA_SRR13565274_GSM5047783_GW21-04_cerebellum_Homo_sapiens_RNA-Seq_kOut</th>\n",
       "      <td>222</td>\n",
       "      <td>40</td>\n",
       "      <td>DMR</td>\n",
       "      <td>RE_Gly</td>\n",
       "      <td>Cb</td>\n",
       "      <td>310 CBX Golgi Gly-Gaba</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTCACAGACGCAA_SRR13565274_GSM5047783_GW21-04_cerebellum_Homo_sapiens_RNA-Seq_kOut</th>\n",
       "      <td>78</td>\n",
       "      <td>22</td>\n",
       "      <td>DMR</td>\n",
       "      <td>RE_EOMES_UBC</td>\n",
       "      <td>Cb</td>\n",
       "      <td>118 ADP-MPO Trp73 Glut</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTCACATAACCTG_SRR13565274_GSM5047783_GW21-04_cerebellum_Homo_sapiens_RNA-Seq_kOut</th>\n",
       "      <td>108</td>\n",
       "      <td>38</td>\n",
       "      <td>P</td>\n",
       "      <td>Progen_OPC</td>\n",
       "      <td>Cb</td>\n",
       "      <td>326 OPC NN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTCATCGGTTCGG_SRR13565274_GSM5047783_GW21-04_cerebellum_Homo_sapiens_RNA-Seq_kOut</th>\n",
       "      <td>73</td>\n",
       "      <td>20</td>\n",
       "      <td>DMR</td>\n",
       "      <td>RE_ATOH1_Granule</td>\n",
       "      <td>Cb</td>\n",
       "      <td>314 CB Granule Glut</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1793824 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                   level_2 level_1  \\\n",
       "TAGTATGTAATA_e14-WT8                                   157      26   \n",
       "TTCGCATAGGTT_e14-WT8                                   271      11   \n",
       "CCTCCAATTGGC_e14-WT8                                   216       5   \n",
       "CACGATATCCAT_e14-WT8                                   373      41   \n",
       "CAGCCGCGCCCT_e14-WT8                                   214       5   \n",
       "...                                                    ...     ...   \n",
       "TTTGGTTTCCCTTGTG_SRR13565274_GSM5047783_GW21-04...     398      29   \n",
       "TTTGGTTTCTGCGGCA_SRR13565274_GSM5047783_GW21-04...     222      40   \n",
       "TTTGTCACAGACGCAA_SRR13565274_GSM5047783_GW21-04...      78      22   \n",
       "TTTGTCACATAACCTG_SRR13565274_GSM5047783_GW21-04...     108      38   \n",
       "TTTGTCATCGGTTCGG_SRR13565274_GSM5047783_GW21-04...      73      20   \n",
       "\n",
       "                                                   new_neighborhood_extended  \\\n",
       "TAGTATGTAATA_e14-WT8                                                       P   \n",
       "TTCGCATAGGTT_e14-WT8                                                       T   \n",
       "CCTCCAATTGGC_e14-WT8                                                       T   \n",
       "CACGATATCCAT_e14-WT8                                                      NN   \n",
       "CAGCCGCGCCCT_e14-WT8                                                       T   \n",
       "...                                                                      ...   \n",
       "TTTGGTTTCCCTTGTG_SRR13565274_GSM5047783_GW21-04...                       DMR   \n",
       "TTTGGTTTCTGCGGCA_SRR13565274_GSM5047783_GW21-04...                       DMR   \n",
       "TTTGTCACAGACGCAA_SRR13565274_GSM5047783_GW21-04...                       DMR   \n",
       "TTTGTCACATAACCTG_SRR13565274_GSM5047783_GW21-04...                         P   \n",
       "TTTGTCATCGGTTCGG_SRR13565274_GSM5047783_GW21-04...                       DMR   \n",
       "\n",
       "                                                        ccl_extended region  \\\n",
       "TAGTATGTAATA_e14-WT8                                             NPC    Ctx   \n",
       "TTCGCATAGGTT_e14-WT8                                          Ctx_IT    Ctx   \n",
       "CCTCCAATTGGC_e14-WT8                                        Ctx_Allo    Ctx   \n",
       "CACGATATCCAT_e14-WT8                                     Endothelial    Ctx   \n",
       "CAGCCGCGCCCT_e14-WT8                                        Ctx_Allo    Ctx   \n",
       "...                                                              ...    ...   \n",
       "TTTGGTTTCCCTTGTG_SRR13565274_GSM5047783_GW21-04...  RE_ATOH1_Granule     Cb   \n",
       "TTTGGTTTCTGCGGCA_SRR13565274_GSM5047783_GW21-04...            RE_Gly     Cb   \n",
       "TTTGTCACAGACGCAA_SRR13565274_GSM5047783_GW21-04...      RE_EOMES_UBC     Cb   \n",
       "TTTGTCACATAACCTG_SRR13565274_GSM5047783_GW21-04...        Progen_OPC     Cb   \n",
       "TTTGTCATCGGTTCGG_SRR13565274_GSM5047783_GW21-04...  RE_ATOH1_Granule     Cb   \n",
       "\n",
       "                                                    abc_subclass_markercor  \n",
       "TAGTATGTAATA_e14-WT8                                      320 Astro-OLF NN  \n",
       "TTCGCATAGGTT_e14-WT8                                   309 CB PLI Gly-Gaba  \n",
       "CCTCCAATTGGC_e14-WT8                                   309 CB PLI Gly-Gaba  \n",
       "CACGATATCCAT_e14-WT8                                            329 ABC NN  \n",
       "CAGCCGCGCCCT_e14-WT8                                   309 CB PLI Gly-Gaba  \n",
       "...                                                                    ...  \n",
       "TTTGGTTTCCCTTGTG_SRR13565274_GSM5047783_GW21-04...     314 CB Granule Glut  \n",
       "TTTGGTTTCTGCGGCA_SRR13565274_GSM5047783_GW21-04...  310 CBX Golgi Gly-Gaba  \n",
       "TTTGTCACAGACGCAA_SRR13565274_GSM5047783_GW21-04...  118 ADP-MPO Trp73 Glut  \n",
       "TTTGTCACATAACCTG_SRR13565274_GSM5047783_GW21-04...              326 OPC NN  \n",
       "TTTGTCATCGGTTCGG_SRR13565274_GSM5047783_GW21-04...     314 CB Granule Glut  \n",
       "\n",
       "[1793824 rows x 6 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs.loc[:,['level_2','level_1','new_neighborhood_extended','ccl_extended','region','abc_subclass_markercor']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "c1c9e027-224a-4434-9d41-910366774841",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['G2M_score', 'S_score', 'batch_name', 'general_region',\n",
       "       'log10_n_counts', 'phase', 'region', 'species', 'female',\n",
       "       'general_region_species', '_scvi_discov_ind', '_scvi_batch_ind',\n",
       "       'psi_0', 'psi_1', 'psi_2', 'q_score', 'level_0', 'level_1', 'level_2',\n",
       "       'antipode_cluster', 'kmeans', 'anno_level_2', 'initial_class',\n",
       "       'ic_extended', 'abc_subclass_markercor', 'juv_subclass_markercor',\n",
       "       'linnarson_extended', 'lamanno_class', 'lamanno_subclass',\n",
       "       'lamanno_cluster', 'neighborhood', 'curated_class_label', 'new_ccl',\n",
       "       'new_neighborhood', 'ccl_extended', 'new_neighborhood_extended',\n",
       "       'n_genes', 'percent_ribo', 'msregion', 'singlet', 'individual',\n",
       "       'timepoint', 'n_counts', 'doublet', 'dataset_name', 'log_abs_sum_dynam',\n",
       "       'clipped_psi_2', 'log_n_genes'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "64fcb5b0-4c3e-4ede-88e8-9bc9606337ab",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{0: 'FOXB1', 1: 'MED29', 2: 'SP8', 6: 'RABIF', 8: 'FZD10-AS1', 9: 'DCTPP1', 11: 'YWHAZ', 13: 'CCT6A', 14: 'TSHZ1', 15: 'SCRT2', 16: 'UBXN1', 19: 'CDC25B', 24: 'SLC16A6', 25: 'TRAPPC2', 26: 'CHST7', 27: 'LUM', 28: 'PRDM8', 31: 'ALYREF', 34: 'CCNO', 35: 'SRM', 36: 'GID8', 39: 'AJUBA', 41: 'CORIN', 42: 'GABRG2', 44: 'CAD', 45: 'INHBA', 47: 'RAD21', 48: 'GNG8', 49: 'NOP58', 50: 'EFNA2', 58: 'PROX1', 59: 'CALCRL', 61: 'SNX7', 62: 'FBLL1', 65: 'MDH1', 66: 'CX3CR1', 67: 'EIF1B', 69: 'SLC6A11', 71: 'TRAPPC2', 73: 'NTF3', 76: 'PDLIM7', 78: 'ZFHX4', 79: 'TMEM132C', 80: 'NBDY', 81: 'OLA1', 82: 'RSPO1', 83: 'TRAPPC2', 85: 'MYBL1', 87: 'WNT3', 88: 'UBE2S', 89: 'EIF3H', 90: 'SOSTDC1', 91: 'GFAP', 94: 'NR2F2', 96: 'ROBO3', 97: 'SERBP1', 98: 'MCIDAS', 100: 'PTCH1', 101: 'PDGFRA', 102: 'MYLK', 103: 'ACSL6', 104: 'ISL1', 108: 'MYC', 109: 'TRAPPC2', 111: 'PDZD11', 112: 'DDX31', 114: 'GAS2', 116: 'PFDN5', 117: 'TRAPPC2', 118: 'PIMREG', 120: 'UBE2V2', 121: 'TP73', 122: 'LMNB1', 123: 'LHX6', 124: 'RACK1', 125: 'HOPX', 126: 'NSFL1C', 127: 'E2F7', 128: 'AIF1', 130: 'TIGAR', 131: 'RGS5', 133: 'PAK2', 134: 'TEX30', 135: 'KANK3', 140: 'ZC3H12C', 143: 'DLL3', 144: 'SERTAD3', 145: 'POMC', 146: 'SEMA7A', 147: 'P2RY1', 151: 'SLU7', 152: 'YPEL5', 155: 'UBE2S', 156: 'SPP1', 157: 'GLI3', 159: 'NR2F2', 160: 'TRAPPC2', 161: 'GNAI3', 164: 'DLL1', 165: 'TRAPPC2', 166: 'MRPS27', 167: 'DLX6', 168: 'TRAPPC2', 172: 'CDKN1C', 174: 'SIX3', 176: 'DLL3', 179: 'SP3', 180: 'LHX5', 181: 'TRAPPC2', 182: 'MBNL3', 183: 'SEMA7A', 184: 'FZD8', 186: 'UBE2S', 187: 'LYVE1', 188: 'TRAPPC2', 190: 'TOX', 194: 'RPL34', 197: 'CMPK1', 198: 'TTC28', 199: 'PDIA6', 200: 'NXPH3', 201: 'FBXO5', 202: 'EGFR', 203: 'HNRNPK', 205: 'TCEAL5', 207: 'TAC1', 208: 'TCERG1L', 209: 'MAFB', 210: 'VCAM1', 213: 'NEUROD4', 214: 'ZBTB20', 215: 'USP4', 216: 'TRPS1', 217: 'NECAB1', 218: 'MTFR1', 219: 'ISL1', 220: 'PECAM1', 221: 'ROBO3', 222: 'ZIC3', 224: 'METTL1', 225: 'FAM136A', 227: 'PTGDS', 228: 'DDX31', 230: 'TAL2', 231: 'MYORG', 234: 'CWC27', 236: 'BCAS1', 238: 'TMEM11', 239: 'EIF1B', 242: 'NR2F2', 243: 'E2F1', 244: 'AURKB', 247: 'GDAP1L1', 249: 'ACTA2', 251: 'TMEM11', 254: 'LMAN1', 256: 'EIF3H', 258: 'KLC2', 259: 'THAP7', 261: 'SP8', 263: 'IL17RD', 264: 'NONO', 265: 'YPEL5', 266: 'NOG', 267: 'NGDN', 269: 'SNCG', 270: 'DNAH9', 271: 'SLF1', 273: 'RSPO3', 275: 'GADD45A', 280: 'TFDP2', 281: 'NEUROG1', 282: 'MAG', 283: 'LHX9', 285: 'IGHM', 287: 'SNCG', 289: 'PROX1', 291: 'SPC24', 293: 'HINT1', 295: 'CCDC153', 296: 'MAPK12', 300: 'UBE2S', 303: 'CORT', 304: 'PEA15', 306: 'NPY', 307: 'BRINP2', 308: 'OLIG1', 313: 'POU3F2', 321: 'RMI2', 324: 'NXPH3', 328: 'GPM6A', 330: 'AGAP2', 333: 'BRINP2', 336: 'ZDHHC6', 337: 'RGS5', 338: 'PAFAH1B2', 339: 'FDPS', 340: 'NR4A1', 341: 'BUB3', 343: 'ZFP64', 344: 'ZIC5', 345: 'BUB3', 346: 'HS3ST4', 349: 'ADAMTS5', 351: 'SPARCL1', 353: 'STMN1', 354: 'E2F1', 356: 'EPOP', 357: 'ANGPTL1', 360: 'ISL1', 361: 'TRAPPC2', 363: 'FANCF', 365: 'FGF3', 367: 'TTR', 368: 'GJA4', 369: 'H1-0', 370: 'H1-1', 373: 'KCNJ8', 374: 'H1-1', 375: 'NRTN', 377: 'EIF4B', 382: 'PPIG', 383: 'MFAP2', 385: 'EVI5', 388: 'ACTR1A', 389: 'CPLX2', 390: 'NEUROG2', 392: 'NRP1', 394: 'CHAC1', 395: 'SQLE', 397: 'POMC', 398: 'MINDY2'}"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "selected_markers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "27ecded3-81d7-4c43-8d90-cdca717c73fc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TRAPPC2</th>\n",
       "      <th>C17orf49</th>\n",
       "      <th>C2orf68</th>\n",
       "      <th>C4orf19</th>\n",
       "      <th>C4orf54</th>\n",
       "      <th>C11orf58</th>\n",
       "      <th>C9orf24</th>\n",
       "      <th>C11orf1</th>\n",
       "      <th>C3orf80</th>\n",
       "      <th>C8orf33</th>\n",
       "      <th>...</th>\n",
       "      <th>ZW10</th>\n",
       "      <th>ZWILCH</th>\n",
       "      <th>ZWINT</th>\n",
       "      <th>ZXDC</th>\n",
       "      <th>ZYG11A</th>\n",
       "      <th>ZYG11B</th>\n",
       "      <th>ZYX</th>\n",
       "      <th>ZZEF1</th>\n",
       "      <th>ASIP</th>\n",
       "      <th>CCDC198</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>-0.372538</td>\n",
       "      <td>-0.253029</td>\n",
       "      <td>-0.666332</td>\n",
       "      <td>-2.484574</td>\n",
       "      <td>-0.566978</td>\n",
       "      <td>-0.116801</td>\n",
       "      <td>-2.204294</td>\n",
       "      <td>-0.524150</td>\n",
       "      <td>-1.046877</td>\n",
       "      <td>-0.494489</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.634458</td>\n",
       "      <td>-1.281075</td>\n",
       "      <td>-1.039370</td>\n",
       "      <td>-0.555655</td>\n",
       "      <td>-0.386801</td>\n",
       "      <td>-0.492599</td>\n",
       "      <td>-1.034181</td>\n",
       "      <td>-0.821047</td>\n",
       "      <td>-1.714158</td>\n",
       "      <td>-1.767640</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>-0.308917</td>\n",
       "      <td>-0.290666</td>\n",
       "      <td>-0.665656</td>\n",
       "      <td>-1.949094</td>\n",
       "      <td>-0.628884</td>\n",
       "      <td>-0.171775</td>\n",
       "      <td>-2.423887</td>\n",
       "      <td>-0.562154</td>\n",
       "      <td>-0.582024</td>\n",
       "      <td>-0.479024</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.503906</td>\n",
       "      <td>-0.970146</td>\n",
       "      <td>-1.051725</td>\n",
       "      <td>-0.502962</td>\n",
       "      <td>-0.460542</td>\n",
       "      <td>-0.462685</td>\n",
       "      <td>-1.134745</td>\n",
       "      <td>-0.742455</td>\n",
       "      <td>-1.313092</td>\n",
       "      <td>-1.262529</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>-0.228588</td>\n",
       "      <td>-0.300823</td>\n",
       "      <td>-0.336561</td>\n",
       "      <td>-2.857441</td>\n",
       "      <td>-0.556463</td>\n",
       "      <td>-0.127702</td>\n",
       "      <td>-3.078704</td>\n",
       "      <td>-0.758076</td>\n",
       "      <td>-1.098588</td>\n",
       "      <td>-0.289459</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.409458</td>\n",
       "      <td>-1.402513</td>\n",
       "      <td>-1.566278</td>\n",
       "      <td>-0.393342</td>\n",
       "      <td>-0.653528</td>\n",
       "      <td>-0.507970</td>\n",
       "      <td>-1.620214</td>\n",
       "      <td>-0.714019</td>\n",
       "      <td>-2.040072</td>\n",
       "      <td>-1.814482</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>-0.402192</td>\n",
       "      <td>-0.233066</td>\n",
       "      <td>-0.583554</td>\n",
       "      <td>-2.046477</td>\n",
       "      <td>-0.656744</td>\n",
       "      <td>-0.298640</td>\n",
       "      <td>-2.114796</td>\n",
       "      <td>-0.663250</td>\n",
       "      <td>-1.106621</td>\n",
       "      <td>-0.393551</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.593605</td>\n",
       "      <td>-0.979246</td>\n",
       "      <td>-1.471757</td>\n",
       "      <td>-0.575624</td>\n",
       "      <td>-0.599983</td>\n",
       "      <td>-0.465162</td>\n",
       "      <td>-1.035159</td>\n",
       "      <td>-0.687485</td>\n",
       "      <td>-1.304722</td>\n",
       "      <td>-1.496459</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>-0.504078</td>\n",
       "      <td>-0.616601</td>\n",
       "      <td>-0.602913</td>\n",
       "      <td>-1.019199</td>\n",
       "      <td>-0.755342</td>\n",
       "      <td>-0.327094</td>\n",
       "      <td>-2.188931</td>\n",
       "      <td>-0.373406</td>\n",
       "      <td>-1.158091</td>\n",
       "      <td>-0.173556</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.600193</td>\n",
       "      <td>-0.995621</td>\n",
       "      <td>-1.769102</td>\n",
       "      <td>-0.505452</td>\n",
       "      <td>-0.932108</td>\n",
       "      <td>-0.761712</td>\n",
       "      <td>-1.082626</td>\n",
       "      <td>-0.811206</td>\n",
       "      <td>-0.973598</td>\n",
       "      <td>-0.647918</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>392</th>\n",
       "      <td>-0.291091</td>\n",
       "      <td>-0.212329</td>\n",
       "      <td>-0.528351</td>\n",
       "      <td>-2.275191</td>\n",
       "      <td>-0.586600</td>\n",
       "      <td>-0.251136</td>\n",
       "      <td>-2.305926</td>\n",
       "      <td>-0.633504</td>\n",
       "      <td>-0.958233</td>\n",
       "      <td>-0.340975</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.565279</td>\n",
       "      <td>-1.098604</td>\n",
       "      <td>-1.445760</td>\n",
       "      <td>-0.512159</td>\n",
       "      <td>-0.537609</td>\n",
       "      <td>-0.418328</td>\n",
       "      <td>-1.506285</td>\n",
       "      <td>-0.690056</td>\n",
       "      <td>-1.497523</td>\n",
       "      <td>-1.722514</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>394</th>\n",
       "      <td>-0.334852</td>\n",
       "      <td>-0.174300</td>\n",
       "      <td>-0.528525</td>\n",
       "      <td>-2.541744</td>\n",
       "      <td>-0.556705</td>\n",
       "      <td>-0.147669</td>\n",
       "      <td>-2.436520</td>\n",
       "      <td>-0.622193</td>\n",
       "      <td>-0.950077</td>\n",
       "      <td>-0.350124</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.507679</td>\n",
       "      <td>-1.144274</td>\n",
       "      <td>-1.129043</td>\n",
       "      <td>-0.479358</td>\n",
       "      <td>-0.555033</td>\n",
       "      <td>-0.489833</td>\n",
       "      <td>-1.280189</td>\n",
       "      <td>-0.691462</td>\n",
       "      <td>-1.580355</td>\n",
       "      <td>-1.528670</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>395</th>\n",
       "      <td>-1.136095</td>\n",
       "      <td>-1.219385</td>\n",
       "      <td>-0.339539</td>\n",
       "      <td>-2.942554</td>\n",
       "      <td>-0.462002</td>\n",
       "      <td>-1.264198</td>\n",
       "      <td>-3.715767</td>\n",
       "      <td>-1.457268</td>\n",
       "      <td>-0.896159</td>\n",
       "      <td>-1.306719</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.244886</td>\n",
       "      <td>-1.412037</td>\n",
       "      <td>-2.020369</td>\n",
       "      <td>-0.229168</td>\n",
       "      <td>-0.216617</td>\n",
       "      <td>-0.183995</td>\n",
       "      <td>-1.811603</td>\n",
       "      <td>-0.026170</td>\n",
       "      <td>-1.869090</td>\n",
       "      <td>-1.828967</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>397</th>\n",
       "      <td>-0.208884</td>\n",
       "      <td>-0.304163</td>\n",
       "      <td>-0.628585</td>\n",
       "      <td>-1.832158</td>\n",
       "      <td>-0.551590</td>\n",
       "      <td>-0.105379</td>\n",
       "      <td>-1.910649</td>\n",
       "      <td>-0.394144</td>\n",
       "      <td>-0.831769</td>\n",
       "      <td>-0.311847</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.498319</td>\n",
       "      <td>-1.000786</td>\n",
       "      <td>-1.088759</td>\n",
       "      <td>-0.562266</td>\n",
       "      <td>-0.395487</td>\n",
       "      <td>-0.526158</td>\n",
       "      <td>-1.261608</td>\n",
       "      <td>-0.748422</td>\n",
       "      <td>-1.646842</td>\n",
       "      <td>-1.537614</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>398</th>\n",
       "      <td>-0.465498</td>\n",
       "      <td>-0.314589</td>\n",
       "      <td>-0.427194</td>\n",
       "      <td>-2.049475</td>\n",
       "      <td>-0.563240</td>\n",
       "      <td>-0.245067</td>\n",
       "      <td>-2.426646</td>\n",
       "      <td>-0.686881</td>\n",
       "      <td>-1.241810</td>\n",
       "      <td>-0.387065</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.451209</td>\n",
       "      <td>-1.031565</td>\n",
       "      <td>-1.723376</td>\n",
       "      <td>-0.319274</td>\n",
       "      <td>-0.639431</td>\n",
       "      <td>-0.393306</td>\n",
       "      <td>-0.901133</td>\n",
       "      <td>-0.449917</td>\n",
       "      <td>-1.589822</td>\n",
       "      <td>-1.676632</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>232 rows × 16738 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      TRAPPC2  C17orf49   C2orf68   C4orf19   C4orf54  C11orf58   C9orf24  \\\n",
       "0   -0.372538 -0.253029 -0.666332 -2.484574 -0.566978 -0.116801 -2.204294   \n",
       "1   -0.308917 -0.290666 -0.665656 -1.949094 -0.628884 -0.171775 -2.423887   \n",
       "2   -0.228588 -0.300823 -0.336561 -2.857441 -0.556463 -0.127702 -3.078704   \n",
       "6   -0.402192 -0.233066 -0.583554 -2.046477 -0.656744 -0.298640 -2.114796   \n",
       "8   -0.504078 -0.616601 -0.602913 -1.019199 -0.755342 -0.327094 -2.188931   \n",
       "..        ...       ...       ...       ...       ...       ...       ...   \n",
       "392 -0.291091 -0.212329 -0.528351 -2.275191 -0.586600 -0.251136 -2.305926   \n",
       "394 -0.334852 -0.174300 -0.528525 -2.541744 -0.556705 -0.147669 -2.436520   \n",
       "395 -1.136095 -1.219385 -0.339539 -2.942554 -0.462002 -1.264198 -3.715767   \n",
       "397 -0.208884 -0.304163 -0.628585 -1.832158 -0.551590 -0.105379 -1.910649   \n",
       "398 -0.465498 -0.314589 -0.427194 -2.049475 -0.563240 -0.245067 -2.426646   \n",
       "\n",
       "      C11orf1   C3orf80   C8orf33  ...      ZW10    ZWILCH     ZWINT  \\\n",
       "0   -0.524150 -1.046877 -0.494489  ... -0.634458 -1.281075 -1.039370   \n",
       "1   -0.562154 -0.582024 -0.479024  ... -0.503906 -0.970146 -1.051725   \n",
       "2   -0.758076 -1.098588 -0.289459  ... -0.409458 -1.402513 -1.566278   \n",
       "6   -0.663250 -1.106621 -0.393551  ... -0.593605 -0.979246 -1.471757   \n",
       "8   -0.373406 -1.158091 -0.173556  ... -0.600193 -0.995621 -1.769102   \n",
       "..        ...       ...       ...  ...       ...       ...       ...   \n",
       "392 -0.633504 -0.958233 -0.340975  ... -0.565279 -1.098604 -1.445760   \n",
       "394 -0.622193 -0.950077 -0.350124  ... -0.507679 -1.144274 -1.129043   \n",
       "395 -1.457268 -0.896159 -1.306719  ... -0.244886 -1.412037 -2.020369   \n",
       "397 -0.394144 -0.831769 -0.311847  ... -0.498319 -1.000786 -1.088759   \n",
       "398 -0.686881 -1.241810 -0.387065  ... -0.451209 -1.031565 -1.723376   \n",
       "\n",
       "         ZXDC    ZYG11A    ZYG11B       ZYX     ZZEF1      ASIP   CCDC198  \n",
       "0   -0.555655 -0.386801 -0.492599 -1.034181 -0.821047 -1.714158 -1.767640  \n",
       "1   -0.502962 -0.460542 -0.462685 -1.134745 -0.742455 -1.313092 -1.262529  \n",
       "2   -0.393342 -0.653528 -0.507970 -1.620214 -0.714019 -2.040072 -1.814482  \n",
       "6   -0.575624 -0.599983 -0.465162 -1.035159 -0.687485 -1.304722 -1.496459  \n",
       "8   -0.505452 -0.932108 -0.761712 -1.082626 -0.811206 -0.973598 -0.647918  \n",
       "..        ...       ...       ...       ...       ...       ...       ...  \n",
       "392 -0.512159 -0.537609 -0.418328 -1.506285 -0.690056 -1.497523 -1.722514  \n",
       "394 -0.479358 -0.555033 -0.489833 -1.280189 -0.691462 -1.580355 -1.528670  \n",
       "395 -0.229168 -0.216617 -0.183995 -1.811603 -0.026170 -1.869090 -1.828967  \n",
       "397 -0.562266 -0.395487 -0.526158 -1.261608 -0.748422 -1.646842 -1.537614  \n",
       "398 -0.319274 -0.639431 -0.393306 -0.901133 -0.449917 -1.589822 -1.676632  \n",
       "\n",
       "[232 rows x 16738 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "coefdf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "8eff0509-96d0-4b33-bc01-0121bd333f60",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TRAPPC2</th>\n",
       "      <th>C17orf49</th>\n",
       "      <th>C2orf68</th>\n",
       "      <th>C4orf19</th>\n",
       "      <th>C4orf54</th>\n",
       "      <th>C11orf58</th>\n",
       "      <th>C9orf24</th>\n",
       "      <th>C11orf1</th>\n",
       "      <th>C3orf80</th>\n",
       "      <th>C8orf33</th>\n",
       "      <th>...</th>\n",
       "      <th>ZW10</th>\n",
       "      <th>ZWILCH</th>\n",
       "      <th>ZWINT</th>\n",
       "      <th>ZXDC</th>\n",
       "      <th>ZYG11A</th>\n",
       "      <th>ZYG11B</th>\n",
       "      <th>ZYX</th>\n",
       "      <th>ZZEF1</th>\n",
       "      <th>ASIP</th>\n",
       "      <th>CCDC198</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Hypo_Glut_4</th>\n",
       "      <td>-0.820873</td>\n",
       "      <td>-1.177970</td>\n",
       "      <td>-1.094952</td>\n",
       "      <td>-5.603813</td>\n",
       "      <td>-2.294325</td>\n",
       "      <td>-0.163394</td>\n",
       "      <td>-6.827879</td>\n",
       "      <td>-0.696181</td>\n",
       "      <td>-2.858164</td>\n",
       "      <td>-0.841368</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.312536</td>\n",
       "      <td>-3.356310</td>\n",
       "      <td>-3.688303</td>\n",
       "      <td>-1.358644</td>\n",
       "      <td>-2.220021</td>\n",
       "      <td>-0.952205</td>\n",
       "      <td>-2.065985</td>\n",
       "      <td>-2.203243</td>\n",
       "      <td>-4.012577</td>\n",
       "      <td>-4.853777</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DE-ME_GABA</th>\n",
       "      <td>-0.366670</td>\n",
       "      <td>-0.749033</td>\n",
       "      <td>-1.157578</td>\n",
       "      <td>-4.547653</td>\n",
       "      <td>-2.225616</td>\n",
       "      <td>-0.276089</td>\n",
       "      <td>-4.737088</td>\n",
       "      <td>-0.979512</td>\n",
       "      <td>-1.535116</td>\n",
       "      <td>-1.084131</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.117136</td>\n",
       "      <td>-3.855460</td>\n",
       "      <td>-4.832050</td>\n",
       "      <td>-1.233255</td>\n",
       "      <td>-1.650649</td>\n",
       "      <td>-0.899614</td>\n",
       "      <td>-1.937440</td>\n",
       "      <td>-1.084174</td>\n",
       "      <td>-4.102415</td>\n",
       "      <td>-3.940735</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LGE_FOXP2_TSHZ1</th>\n",
       "      <td>-0.383610</td>\n",
       "      <td>-0.992511</td>\n",
       "      <td>-0.615233</td>\n",
       "      <td>-5.260323</td>\n",
       "      <td>-1.343347</td>\n",
       "      <td>-0.434168</td>\n",
       "      <td>-6.165697</td>\n",
       "      <td>-2.145880</td>\n",
       "      <td>-4.061696</td>\n",
       "      <td>-0.442142</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.060038</td>\n",
       "      <td>-2.465349</td>\n",
       "      <td>-5.369750</td>\n",
       "      <td>-0.286909</td>\n",
       "      <td>-2.178013</td>\n",
       "      <td>-0.733180</td>\n",
       "      <td>-2.498071</td>\n",
       "      <td>-0.698002</td>\n",
       "      <td>-3.404821</td>\n",
       "      <td>-4.246020</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ctx_IT</th>\n",
       "      <td>-0.594849</td>\n",
       "      <td>-0.804239</td>\n",
       "      <td>-0.738485</td>\n",
       "      <td>-5.805456</td>\n",
       "      <td>-1.333528</td>\n",
       "      <td>-0.458989</td>\n",
       "      <td>-6.558375</td>\n",
       "      <td>-1.233735</td>\n",
       "      <td>-1.982386</td>\n",
       "      <td>-0.536484</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.369197</td>\n",
       "      <td>-3.195014</td>\n",
       "      <td>-5.432046</td>\n",
       "      <td>-0.700736</td>\n",
       "      <td>-1.955970</td>\n",
       "      <td>-0.626115</td>\n",
       "      <td>-2.375831</td>\n",
       "      <td>-0.896224</td>\n",
       "      <td>-3.815291</td>\n",
       "      <td>-4.493143</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Progen_Astro-Ependymal</th>\n",
       "      <td>-0.923899</td>\n",
       "      <td>-1.082071</td>\n",
       "      <td>-1.177198</td>\n",
       "      <td>-1.605571</td>\n",
       "      <td>-2.317445</td>\n",
       "      <td>-0.704928</td>\n",
       "      <td>-1.591669</td>\n",
       "      <td>-0.730893</td>\n",
       "      <td>-3.498423</td>\n",
       "      <td>-0.204062</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.314128</td>\n",
       "      <td>-1.684216</td>\n",
       "      <td>-3.020839</td>\n",
       "      <td>-0.848463</td>\n",
       "      <td>-2.210559</td>\n",
       "      <td>-1.894555</td>\n",
       "      <td>-1.671324</td>\n",
       "      <td>-1.549646</td>\n",
       "      <td>-4.035697</td>\n",
       "      <td>-2.440551</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Hypo_Glut_2</th>\n",
       "      <td>-0.926633</td>\n",
       "      <td>-1.326576</td>\n",
       "      <td>-1.275196</td>\n",
       "      <td>-5.031530</td>\n",
       "      <td>-2.232918</td>\n",
       "      <td>-0.474176</td>\n",
       "      <td>-6.734744</td>\n",
       "      <td>-1.846870</td>\n",
       "      <td>-3.440171</td>\n",
       "      <td>-0.440405</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.081448</td>\n",
       "      <td>-3.447381</td>\n",
       "      <td>-7.262494</td>\n",
       "      <td>-0.932041</td>\n",
       "      <td>-1.961726</td>\n",
       "      <td>-0.864031</td>\n",
       "      <td>-1.866808</td>\n",
       "      <td>-1.333600</td>\n",
       "      <td>-3.951169</td>\n",
       "      <td>-4.792369</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RE_Gaba_Purkinje</th>\n",
       "      <td>-0.453021</td>\n",
       "      <td>-0.448345</td>\n",
       "      <td>-2.886758</td>\n",
       "      <td>-6.339574</td>\n",
       "      <td>-1.889843</td>\n",
       "      <td>-0.589934</td>\n",
       "      <td>-6.360331</td>\n",
       "      <td>-1.301616</td>\n",
       "      <td>-2.816903</td>\n",
       "      <td>-1.019424</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.773852</td>\n",
       "      <td>-4.274704</td>\n",
       "      <td>-3.943422</td>\n",
       "      <td>-0.956561</td>\n",
       "      <td>-2.803904</td>\n",
       "      <td>-1.369598</td>\n",
       "      <td>-2.704681</td>\n",
       "      <td>-1.784678</td>\n",
       "      <td>-4.260201</td>\n",
       "      <td>-4.218660</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MGE_CRABP1_MAF_GABA</th>\n",
       "      <td>-0.714993</td>\n",
       "      <td>-1.054392</td>\n",
       "      <td>-0.615061</td>\n",
       "      <td>-5.687375</td>\n",
       "      <td>-1.889750</td>\n",
       "      <td>-0.420267</td>\n",
       "      <td>-7.077633</td>\n",
       "      <td>-1.908258</td>\n",
       "      <td>-4.264877</td>\n",
       "      <td>-0.666494</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.159883</td>\n",
       "      <td>-3.354592</td>\n",
       "      <td>-5.513024</td>\n",
       "      <td>-0.875563</td>\n",
       "      <td>-2.215279</td>\n",
       "      <td>-0.823422</td>\n",
       "      <td>-3.165290</td>\n",
       "      <td>-1.018773</td>\n",
       "      <td>-3.608001</td>\n",
       "      <td>-4.449201</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Hypo_Glut_1</th>\n",
       "      <td>-0.249886</td>\n",
       "      <td>-0.925681</td>\n",
       "      <td>-1.110769</td>\n",
       "      <td>-5.157996</td>\n",
       "      <td>-2.031756</td>\n",
       "      <td>-0.225042</td>\n",
       "      <td>-8.171231</td>\n",
       "      <td>-0.965551</td>\n",
       "      <td>-4.734511</td>\n",
       "      <td>-0.831364</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.372312</td>\n",
       "      <td>-3.431541</td>\n",
       "      <td>-4.754654</td>\n",
       "      <td>-1.150761</td>\n",
       "      <td>-3.767961</td>\n",
       "      <td>-1.199426</td>\n",
       "      <td>-2.256195</td>\n",
       "      <td>-1.517096</td>\n",
       "      <td>-4.077635</td>\n",
       "      <td>-4.918835</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DE-ME_GABA_Dopa</th>\n",
       "      <td>-0.217104</td>\n",
       "      <td>-0.894106</td>\n",
       "      <td>-1.120704</td>\n",
       "      <td>-4.618561</td>\n",
       "      <td>-2.373845</td>\n",
       "      <td>-0.124338</td>\n",
       "      <td>-5.292165</td>\n",
       "      <td>-0.911878</td>\n",
       "      <td>-2.725606</td>\n",
       "      <td>-0.809679</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.057848</td>\n",
       "      <td>-4.289969</td>\n",
       "      <td>-4.505805</td>\n",
       "      <td>-1.983566</td>\n",
       "      <td>-2.102653</td>\n",
       "      <td>-0.736605</td>\n",
       "      <td>-2.498643</td>\n",
       "      <td>-1.680955</td>\n",
       "      <td>-4.092096</td>\n",
       "      <td>-4.050555</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>67 rows × 16738 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                         TRAPPC2  C17orf49   C2orf68   C4orf19   C4orf54  \\\n",
       "Hypo_Glut_4            -0.820873 -1.177970 -1.094952 -5.603813 -2.294325   \n",
       "DE-ME_GABA             -0.366670 -0.749033 -1.157578 -4.547653 -2.225616   \n",
       "LGE_FOXP2_TSHZ1        -0.383610 -0.992511 -0.615233 -5.260323 -1.343347   \n",
       "Ctx_IT                 -0.594849 -0.804239 -0.738485 -5.805456 -1.333528   \n",
       "Progen_Astro-Ependymal -0.923899 -1.082071 -1.177198 -1.605571 -2.317445   \n",
       "...                          ...       ...       ...       ...       ...   \n",
       "Hypo_Glut_2            -0.926633 -1.326576 -1.275196 -5.031530 -2.232918   \n",
       "RE_Gaba_Purkinje       -0.453021 -0.448345 -2.886758 -6.339574 -1.889843   \n",
       "MGE_CRABP1_MAF_GABA    -0.714993 -1.054392 -0.615061 -5.687375 -1.889750   \n",
       "Hypo_Glut_1            -0.249886 -0.925681 -1.110769 -5.157996 -2.031756   \n",
       "DE-ME_GABA_Dopa        -0.217104 -0.894106 -1.120704 -4.618561 -2.373845   \n",
       "\n",
       "                        C11orf58   C9orf24   C11orf1   C3orf80   C8orf33  ...  \\\n",
       "Hypo_Glut_4            -0.163394 -6.827879 -0.696181 -2.858164 -0.841368  ...   \n",
       "DE-ME_GABA             -0.276089 -4.737088 -0.979512 -1.535116 -1.084131  ...   \n",
       "LGE_FOXP2_TSHZ1        -0.434168 -6.165697 -2.145880 -4.061696 -0.442142  ...   \n",
       "Ctx_IT                 -0.458989 -6.558375 -1.233735 -1.982386 -0.536484  ...   \n",
       "Progen_Astro-Ependymal -0.704928 -1.591669 -0.730893 -3.498423 -0.204062  ...   \n",
       "...                          ...       ...       ...       ...       ...  ...   \n",
       "Hypo_Glut_2            -0.474176 -6.734744 -1.846870 -3.440171 -0.440405  ...   \n",
       "RE_Gaba_Purkinje       -0.589934 -6.360331 -1.301616 -2.816903 -1.019424  ...   \n",
       "MGE_CRABP1_MAF_GABA    -0.420267 -7.077633 -1.908258 -4.264877 -0.666494  ...   \n",
       "Hypo_Glut_1            -0.225042 -8.171231 -0.965551 -4.734511 -0.831364  ...   \n",
       "DE-ME_GABA_Dopa        -0.124338 -5.292165 -0.911878 -2.725606 -0.809679  ...   \n",
       "\n",
       "                            ZW10    ZWILCH     ZWINT      ZXDC    ZYG11A  \\\n",
       "Hypo_Glut_4            -1.312536 -3.356310 -3.688303 -1.358644 -2.220021   \n",
       "DE-ME_GABA             -1.117136 -3.855460 -4.832050 -1.233255 -1.650649   \n",
       "LGE_FOXP2_TSHZ1        -1.060038 -2.465349 -5.369750 -0.286909 -2.178013   \n",
       "Ctx_IT                 -1.369197 -3.195014 -5.432046 -0.700736 -1.955970   \n",
       "Progen_Astro-Ependymal -1.314128 -1.684216 -3.020839 -0.848463 -2.210559   \n",
       "...                          ...       ...       ...       ...       ...   \n",
       "Hypo_Glut_2            -1.081448 -3.447381 -7.262494 -0.932041 -1.961726   \n",
       "RE_Gaba_Purkinje       -1.773852 -4.274704 -3.943422 -0.956561 -2.803904   \n",
       "MGE_CRABP1_MAF_GABA    -1.159883 -3.354592 -5.513024 -0.875563 -2.215279   \n",
       "Hypo_Glut_1            -1.372312 -3.431541 -4.754654 -1.150761 -3.767961   \n",
       "DE-ME_GABA_Dopa        -1.057848 -4.289969 -4.505805 -1.983566 -2.102653   \n",
       "\n",
       "                          ZYG11B       ZYX     ZZEF1      ASIP   CCDC198  \n",
       "Hypo_Glut_4            -0.952205 -2.065985 -2.203243 -4.012577 -4.853777  \n",
       "DE-ME_GABA             -0.899614 -1.937440 -1.084174 -4.102415 -3.940735  \n",
       "LGE_FOXP2_TSHZ1        -0.733180 -2.498071 -0.698002 -3.404821 -4.246020  \n",
       "Ctx_IT                 -0.626115 -2.375831 -0.896224 -3.815291 -4.493143  \n",
       "Progen_Astro-Ependymal -1.894555 -1.671324 -1.549646 -4.035697 -2.440551  \n",
       "...                          ...       ...       ...       ...       ...  \n",
       "Hypo_Glut_2            -0.864031 -1.866808 -1.333600 -3.951169 -4.792369  \n",
       "RE_Gaba_Purkinje       -1.369598 -2.704681 -1.784678 -4.260201 -4.218660  \n",
       "MGE_CRABP1_MAF_GABA    -0.823422 -3.165290 -1.018773 -3.608001 -4.449201  \n",
       "Hypo_Glut_1            -1.199426 -2.256195 -1.517096 -4.077635 -4.918835  \n",
       "DE-ME_GABA_Dopa        -0.736605 -2.498643 -1.680955 -4.092096 -4.050555  \n",
       "\n",
       "[67 rows x 16738 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ccl_coefdf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "ec2e5378-3b72-4a69-bd28-c0d8c3c95458",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/scratch/fast/50132/ipykernel_731618/1481285248.py:16: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  grouped = adata.obs.groupby('level_2')\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "     level_2  level_1 new_neighborhood_extended ccl_extended  \\\n",
      "66       114       19                        NN          ABC   \n",
      "136      220       28                        NN   Angioblast   \n",
      "144      231        6                        NN        Astro   \n",
      "208      357        9                         P        Astro   \n",
      "14        26        6                        NN        Astro   \n",
      "..       ...      ...                       ...          ...   \n",
      "173      283        1                       DMR    Thal_Glut   \n",
      "42        79       19                        NN         VLMC   \n",
      "141      227       19                        NN         VLMC   \n",
      "64       111        6                        NN   unassigned   \n",
      "97       161        6                        NN   unassigned   \n",
      "\n",
      "            abc_subclass_markercor  n_cells  \\\n",
      "66                      329 ABC NN     6891   \n",
      "136                    333 Endo NN     5040   \n",
      "144                319 Astro-TE NN     6171   \n",
      "208                316 Bergmann NN     4798   \n",
      "14                 316 Bergmann NN     9578   \n",
      "..                             ...      ...   \n",
      "173  150 CM-IAD-CL-PCN Sema5b Glut     6591   \n",
      "42                     330 VLMC NN     6601   \n",
      "141                     329 ABC NN     2268   \n",
      "64                 316 Bergmann NN     5193   \n",
      "97                 316 Bergmann NN     8291   \n",
      "\n",
      "                                      top_regions  \\\n",
      "66       Hypo (45.6%), Choroid (21.5%), FB (9.3%)   \n",
      "136            FB (17.3%), HB (17.1%), MB (16.7%)   \n",
      "144  Ctx-Ins (17.3%), Ctx-Hip (15.8%), MB (11.5%)   \n",
      "208          Cb (37.2%), HB (18.9%), Hypo (11.8%)   \n",
      "14           MB (20.6%), HB (15.8%), Thal (11.2%)   \n",
      "..                                            ...   \n",
      "173     Thal (46.0%), FB (22.8%), Ctx-Occ (17.7%)   \n",
      "42             HB (29.3%), FB (18.0%), MB (16.7%)   \n",
      "141            HB (25.7%), MB (19.4%), FB (15.3%)   \n",
      "64         Hypo (24.3%), Cb (16.3%), Thal (14.5%)   \n",
      "97      Cb (31.1%), Hypo (14.7%), Ctx-Hip (11.8%)   \n",
      "\n",
      "                              top5_markers_level2  \\\n",
      "66                   POSTN, SIX2, PAX9, DPT, GAS2   \n",
      "136           NOS2, BCL6B, PCDH12, ADGRL4, PECAM1   \n",
      "144         MYORG, SLC6A11, SLC4A4, CPT1A, ACSBG1   \n",
      "208        PTF1A, ANGPTL1, OR56A4, OR5B21, MIR23B   \n",
      "14          CHST7, IGFBP2, MIR3065, MIR215, PDHA2   \n",
      "..                                            ...   \n",
      "173                 LHX9, GBX2, NRAP, PTH2, KITLG   \n",
      "42         DPT, SLC6A13, COL6A3, CYP1B1, TMEM132C   \n",
      "141  SERPIND1, SLC6A20, SLC22A2, SLC6A13, SLC22A6   \n",
      "64    TKTL1, MIR3065, MIR215, MIRLET7A2, CSN1S2AP   \n",
      "97    CSN1S2AP, MIR215, MIRLET7A2, OR13C8, MIR23B   \n",
      "\n",
      "                              top5_markers_ccl  \n",
      "66             SIX2, TCF21, PAX9, PRRX2, FOXD2  \n",
      "136      ADGRL4, SLC38A5, ROBO4, PECAM1, SOX18  \n",
      "144         SLC6A11, DBX2, AQP4, NCAN, SLC7A10  \n",
      "208         SLC6A11, DBX2, AQP4, NCAN, SLC7A10  \n",
      "14          SLC6A11, DBX2, AQP4, NCAN, SLC7A10  \n",
      "..                                         ...  \n",
      "173           TCF7L2, GBX2, KITLG, NTNG1, PTH2  \n",
      "42   SERPIND1, SLC22A6, PTGDS, CYP1B1, SLC22A2  \n",
      "141  SERPIND1, SLC22A6, PTGDS, CYP1B1, SLC22A2  \n",
      "64         CYP26B1, KLHL25, GDF10, AAMP, DARS1  \n",
      "97         CYP26B1, KLHL25, GDF10, AAMP, DARS1  \n",
      "\n",
      "[232 rows x 9 columns]\n"
     ]
    }
   ],
   "source": [
    "# Function to get mode value (first mode if there are multiple)\n",
    "def get_mode(series):\n",
    "    return series.mode().iloc[0] if not series.mode().empty else np.nan\n",
    "\n",
    "# Function to return top n marker genes from a row (sorted descending by coefficient)\n",
    "def get_top_n_markers(row, n=5):\n",
    "    return list(row.sort_values(ascending=False).head(n).index)\n",
    "\n",
    "# Function to calculate top n regions with proportions formatted as \"region (prop%)\"\n",
    "def get_top_regions(series, n=3):\n",
    "    counts = series.value_counts(normalize=True)\n",
    "    top_n = counts.head(n)\n",
    "    return \", \".join([f\"{idx} ({prop:.1%})\" for idx, prop in top_n.items()])\n",
    "\n",
    "# Group obs by level_2\n",
    "grouped = adata.obs.groupby('level_2')\n",
    "vcs = adata.obs['level_2'].value_counts()\n",
    "results = []\n",
    "for level2, group in grouped:\n",
    "    mode_level_1 = get_mode(group['level_1'])\n",
    "    mode_new_neighborhood = get_mode(group['new_neighborhood_extended'])\n",
    "    mode_ccl = get_mode(group['ccl_extended'])\n",
    "    mode_abc = get_mode(group['abc_subclass_markercor'])\n",
    "    top_regions = get_top_regions(group['region'], n=3)\n",
    "    \n",
    "    # Top 5 markers for level_2 from gene_markers\n",
    "    if level2 in coefdf.index:\n",
    "        top5_level2 = get_top_n_markers(coefdf.loc[level2])\n",
    "    else:\n",
    "        top5_level2 = []\n",
    "    \n",
    "    # Top 5 markers for the level_2's assigned ccl_extended from ccl_markers\n",
    "    if mode_ccl in ccl_coefdf.index:\n",
    "        top5_ccl = get_top_n_markers(ccl_coefdf.loc[mode_ccl])\n",
    "    else:\n",
    "        top5_ccl = []\n",
    "    \n",
    "    results.append({\n",
    "        'level_2': level2,\n",
    "        'level_1': mode_level_1,\n",
    "        'new_neighborhood_extended': mode_new_neighborhood,\n",
    "        'ccl_extended': mode_ccl,\n",
    "        'abc_subclass_markercor': mode_abc,\n",
    "        'n_cells':vcs[level2],\n",
    "        'top_regions': top_regions,\n",
    "        'top5_markers_level2': \", \".join(top5_level2),\n",
    "        'top5_markers_ccl': \", \".join(top5_ccl)\n",
    "    })\n",
    "\n",
    "result_df = pd.DataFrame(results)\n",
    "result_df = result_df.sort_values('ccl_extended')  # order by ccl_extended alphabetically\n",
    "\n",
    "print(result_df)\n",
    "result_df.to_csv(os.path.join(sc.settings.figdir,'cluster_metadata.csv'),sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "c9b7d707-b090-43fb-8497-7a031b2c20bf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "398"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "level2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "c33c09fe-1a42-422b-b751-5ac268c6ab1f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "level_2\n",
       "265    55256\n",
       "36     40998\n",
       "224    38513\n",
       "356    33600\n",
       "35     33163\n",
       "       ...  \n",
       "289      670\n",
       "127      611\n",
       "101      566\n",
       "199      559\n",
       "307      368\n",
       "Name: count, Length: 232, dtype: int64"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs['level_2'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dfff2895-ec36-44e8-8806-aea91dbb5642",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "rapids-24.12",
   "language": "python",
   "name": "rapids-24.12"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
